q2_longitudinal/tests/test_longitudinal.py

```diff
@@ -288,6 +310,149 @@ def test_nmit_missing_table_ids(self):
+            first_differences(
+                metadata=single_sam, state_column='Time',
+                individual_id_column='ind',
+                metric='Value', replicate_handling='drop')
+
+    def test_first_differences_empty(self):
+        with self.assertRaisesRegex(ValueError, "Metadata is empty"):
+            first_differences(
+                metadata=(qiime2.Metadata(pd.DataFrame({'a': [], 'b': []}))),
+                state_column='Time', individual_id_column='ind',
+                metric='Value', replicate_handling='drop')
+
+    def test_first_differences_taxa(self):
+        first_differences(
+            metadata=self.md_ecam_fp, state_column='month',
+            individual_id_column='studyid',
+            metric='e2c3ff4f647112723741aa72087f1bfa',
+            replicate_handling='drop', table=self.table_ecam_fp)
+
+    def test_first_differences_one_subject_many_times(self):
+        exp = pd.Series(
+            [-0.01, 0.01, 0.01, 0.06, -0.01, 0.03, -0.02, 0.03, 0.05, -0.03,
+             0.04],
+            index=['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11'],
+            name='Difference')
+        exp.index.name = '#SampleID'
+        obs = first_differences(
+            metadata=qiime2.Metadata(md_one_subject_many_times),
+            state_column='Time', individual_id_column='ind',
+            metric='Value', replicate_handling='drop')
+        pdt.assert_series_equal(obs, exp)
+
+    def test_first_distances(self):
+        exp = pd.Series([0.1, 0.1, 0.3, 0.1, 0.2, 0.4],
+                        index=['3', '4', '5', '9', '10', '11'],
+                        name='Distance')
+        exp.index.name = '#SampleID'
+        obs = first_distances(
+            distance_matrix=dm, metadata=qiime2.Metadata(md),
+            state_column='Time', individual_id_column='ind',
+            replicate_handling='drop')
+        pdt.assert_series_equal(obs, exp)
+
+    def test_first_distances_single_sample(self):
+        with self.assertRaisesRegex(RuntimeError, "Output is empty"):
+            obs = first_distances(
+                distance_matrix=dm_single_sample, metadata=qiime2.Metadata(md),
+                state_column='Time', individual_id_column='ind',
+                replicate_handling='drop')
+            print(obs)
+
+    def test_first_distances_one_subject_many_times(self):
+        exp = pd.Series(
+            [0.3, 0.9, 0.3, 0.2, 0.4, 0.4, 0.5, 0.8, 0.3, 0.4, 0.6],
+            index=['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11'],
+            name='Distance')
+        exp.index.name = '#SampleID'
+        obs = first_distances(
+            distance_matrix=dm,
+            metadata=qiime2.Metadata(md_one_subject_many_times),
+            state_column='Time', individual_id_column='ind',
+            replicate_handling='drop')
+        pdt.assert_series_equal(obs, exp)
+
+    # just check that it works on real data
```

Comment:
Can you assert the results somehow in this test?

Reply:
👍 I've added the expected results as a tsv